File: Iteration_2/src/main/java/AdvisorInterface.java
import java.util.InputMismatchException;
import java.util.Scanner;

public class AdvisorInterface implements UserInterface {
    Advisor advisor;
    Scanner scan;

    public AdvisorInterface() {
        scan = new Scanner(System.in);
    }

    public AdvisorInterface(Advisor advisor) {
        this.advisor = advisor;
        scan = new Scanner(System.in);
    }

    @Override
    public boolean showMenu() {
        boolean logOut = false;
        switch (getChoice()) {
            case 1:
                logOut = showStudentsMenu();
                break;
            case 2:
                System.out.println("You have successfully logged out\n");
                logOut = true;
                break;
            default:
                System.out.println("Enter 1 or 2.");
        }
        return logOut;
    }

    public int getChoice() {
        System.out.print("1-  Students Menu\n2-  Log Out\nSelect an operation: ");
        int choice = 0;
        try{
        choice = scan.nextInt();
        }catch(InputMismatchException e){
            System.out.println("Enter an integer value.");
        }
        return choice;
    }

    public boolean showStudentsMenu() {
        boolean logOut = false;
        int numberOfStudents = advisor.getStudents().size();
        for (int i = 0; i < numberOfStudents; i++) {
            System.out.println((i + 1) + "-    " + advisor.getStudents().get(i).getID());
        }
        int studentIndex = -1;
        try{
        System.out.print("Which student do you select? :\nIf you want to log out, enter \"0\": ");
        studentIndex = scan.nextInt() - 1;
        }catch(InputMismatchException e){
            System.out.println("Enter an integer value.");
            return false;
        }
        if (studentIndex == -1)
            return false;
        else if (studentIndex >= (numberOfStudents - 1)) {
            System.out.println("Please write a number between 1 and " + numberOfStudents + ".");
            return false;
        }
        Student currentStudent = advisor.getStudents().get(studentIndex);
        if (currentStudent == null)
            System.out.println("Student is null.");

        studentOperations(currentStudent);
        return logOut;
    }

    public void approveCourse(Student student, CourseSection courseSection) {
        advisor.approveCourse(student, courseSection);
        System.out.println(courseSection.getParentCourse().getCourseId() + " has been approved.");
        if (student.getTranscript().getWaitedCourses().isEmpty())
            System.out.println("No more courses to approve.");
    }

    public void rejectCourse(Student student, CourseSection course) {
        advisor.rejectCourse(student, course.getParentCourse());
        System.out.println("The course has been rejected.");
    }

    public boolean courseOperations(Student student, int courseIndex) {
        boolean logOut = false;
    
        // Get Course Section.
        CourseSection courseSection = student.getTranscript().getWaitedCourses().get(courseIndex - 1).getCourseSections().get(0);
    
        // Advisor will check that is there any conflict. If there is, request will be rejected.
        if (!advisor.checkSectionConflict(student, courseSection)) {
            System.out.println("There is a conflict between sections. The course cannot be approved! It is automatically rejected.");
            rejectCourse(student, courseSection);
            return logOut;
        }
    
        // If there isn't any conflict, get confirmation from user.
        System.out.print("Do you want to approve this course?(y/n): \nIf you want to turn back enter \"0\": ");
        String approve = scan.next();
    
        // Check confirmation message.
        if (!approve.equals("0") && !approve.equals("y") && !approve.equals("n")) {
            System.out.println("Enter \"y\", \"n\" or \"0\".");
        } else if (approve.equals("0")) {
            logOut = true;
        } else if (approve.equals("y")) {
            approveCourse(student, courseSection); // Approve Course.
        } else {
            rejectCourse(student, courseSection); // Reject Course.
        }
    
        return logOut;
    }
    

    public void studentOperations(Student student) {
        while (true) {
            if (student.getTranscript().getWaitedCourses().isEmpty()) {
                System.out.println("All waited courses have been approved. You will be directed to main menu.");
                break;
            }
            student.getTranscript().showWaitedCourses();
            int courseIndex = -1;
            System.out.print("Which course do you want  to select?: \nIf you want to turn back enter \"0\": ");
            try {
                courseIndex = scan.nextInt();
                if (courseIndex == 0)
                    break;
                int size = student.getTranscript().getWaitedCourses().size();
                if (courseIndex <= 0 || courseIndex > size) {
                    System.out.println("Enter a value between 1 and " + size);
                } else {
                    if (courseOperations(student, courseIndex))
                        break;
                }
            } catch (InputMismatchException e) {
                System.out.println("Enter an integer value");
                break;
            }

        }
    }
}

File: Iteration_3/AdvisorInterface.py
from NotificationSystem import NotificationSystem
from UserInterface import UserInterface

class AdvisorInterface(UserInterface):
    def __init__(self, advisor : 'Advisor', notification_system: NotificationSystem):
        self.advisor = advisor
        self.scanner = None
        self.notification_system = notification_system

    def show_menu(self):
        log_out = False
        choice = self.get_choice()
        if choice == 1:
            log_out = self.show_students_menu()
        elif choice == 2:
            print("You have successfully logged out\n")
            log_out = True
        else:
            print("Enter 1 or 2.")
        return log_out

    def get_choice(self):
        print("1- Students Menu\n2- Log Out\nSelect an operation: ", end="")
        try:
            choice = int(input())
            return choice
        except ValueError:
            print("Enter an integer value.")
            return 0

    def show_students_menu(self):
        log_out = False
        students = self.advisor.get_students()
        number_of_students = len(students)

        for i, student in enumerate(students):
            print(f"{i + 1}- {student.get_id()}")

        try:
            student_index = int(input("Which student do you select? :\nIf you want to log out, enter \"0\": ")) - 1
        except ValueError:
            print("Enter an integer value.")
            return False

        if student_index == -1:
            return False
        elif student_index >= number_of_students or student_index < 0:
            print(f"Please write a number between 1 and {number_of_students}.")
            return False

        current_student = students[student_index]
        if current_student is None:
            print("Student is null.")

        self.student_operations(current_student)
        return log_out

    def approve_course(self, student, course_section):
        try:
            self.advisor.approve_course(student, course_section)
            self.notification_system.create_notification(sender=self, receiver=student, message="Your registeration to " + course_section.get_section_id() + " has been approved.")
            if not student.get_transcript().get_waited_courses():
                print("No more courses to approve.")
        except Exception as e:
            print(str(e))

    def reject_course(self, student, course_section):
        try:
            self.advisor.reject_course(student, course_section)
            self.notification_system.create_notification(sender=self, receiver=student, message="Your registeration to " + course_section.get_section_id() + " has been rejected.")
        except Exception as e:
            print(str(e))

    def course_operations(self, student, course_index):
        try:
            log_out = False
            # Get Course Section
            parent_course = student.get_transcript().get_waited_courses()[course_index - 1]
            course_section = None

            for section in student.get_transcript().get_waited_sections():
                if section.get_parent_course().get_course_id() == parent_course.get_course_id():
                    course_section = section

            # Check for section conflict
            if not self.advisor.check_section_conflict(student, course_section):
                print("There is a conflict between sections. The course cannot be approved! It is automatically rejected.")
                self.reject_course(student, course_section)
                return log_out

            # User confirmation
            print("Do you want to approve this course? (y/n): \nIf you want to turn back enter \"0\": ", end="")
            approve = input().strip()

            if approve not in ("0", "y", "n"):
                print("Enter \"y\", \"n\" or \"0\".")
            elif approve == "0":
                log_out = True
            elif approve == "y":
                self.approve_course(student, course_section)
            else:
                self.reject_course(student, course_section)

            return log_out

        except Exception as e:
            print(str(e))

    def student_operations(self, student):
        try:
            while True:
                if not student.get_transcript().get_waited_courses():
                    print("All waited courses have been approved. You will be directed to main menu.")
                    break

                student.get_transcript().show_waited_courses()
                try:
                    course_index = int(input("Which course do you want to select?: \nIf you want to turn back enter \"0\": "))
                    if course_index == 0:
                        break
                    size = len(student.get_transcript().get_waited_courses())
                    if course_index <= 0 or course_index > size:
                        print(f"Enter a value between 1 and {size}")
                    else:
                        if self.course_operations(student, course_index):
                            break
                except ValueError:
                    print("Enter an integer value.")
                    break

        except Exception as e:
            print(str(e))